rust/src/core.rs

```diff
@@ -39,6 +39,35 @@ pub const STREAM_GAP:      u8 = 0x10;
 pub const STREAM_DEPTH:    u8 = 0x20;
 pub const STREAM_MIDSTREAM:u8 = 0x40;
 
+#[repr(C)]
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Direction {
+    ToServer = 0x04,
+    ToClient = 0x08,
+}
+
+impl Default for Direction {
+    fn default() -> Self { Direction::ToServer }
+}
+
+impl From<u8> for Direction {
+    fn from(d: u8) -> Self {
+        match d {
+            0x04 => { Direction::ToServer },
+            0x08 => { Direction::ToClient },
+            _ => {
+                if d & 0b0000_0100 != 0 {
+                    Direction::ToServer
+                } else if d & 0b0000_1000 != 0 {
+                    Direction::ToClient
+                } else {
+                    panic!("Unknown direction!!");
```

Comment:
Convert to a TryFrom.  The caller can then decide to panic?  I'm not sure how intrusive this would be, but I don't think an impl from From should panic.

Reply:
I think this could be done using `debug_validate_bug_on`. We want to catch improper uses of the direction, but I don't think we need to do dynamic error checking on every path

Reply:
> @victorjulien If debug validation is not enabled would it be OK to default to `ToServer`? Otherwise there is probably no other option but to panic. If we did use a `Result` for the return type, what would be caller due on err? Probably panic.

This is the solution I'd want. In `debug_validation` mode panic, otherwise default to toserver